Firebase - Case Study (#7)
NoroffFEU/development-platforms: firebase-case-study.md

---
title: Firebase Case Study
keywords: development platforms, firebase, case-study
tags: Development Platforms, Firebase, Case Study

> sidebar: development-platforms/firebase

Reviewer: ```suggestion
sidebar: development-platforms
```

> permalink: development-platforms/firebase-case-study.md

Reviewer: ```suggestion
permalink: development-platforms/case-study-firebase.md
```

> folder: development-platforms/Firebase

Reviewer: ```suggestion
folder: development-platforms
```

---

> ### Table of contents

Reviewer: This is generated automatically from H2s - you can remove it

1. Introduction
2. Brief History
3. Features
4. Strengths
5. Weaknesses
6. Reviews
7. Alternatives & Comparison
8. Summary
9. Credits
10. Companies using Firebase
11. Reference links
12. Video links

____________________________________________________________________________________________________________________

## **Introduction:**

## *What is ***FIREBASE***?*

![Firebase](images/firebase.jpeg)

A mobile development platform with features that help both developers work efficiently, fast, and reliably, solving the three main issues in mobile apps and all multi-platform devices. It was a backend platform that enhanced the developer's reliance on using it based on demand. It is a leading platform that is relied on by developers globally. It has rapid improvements and additional features that will give satisfying results for the knowledge of both developers and users. And since it is cloud-hosted, it will provide a smoother performance for demand scaling without any hassles. All features are functions for both mobile and web-based applications, including services for building, testing, and managing apps.

> - [links](https://blog.back4app.com/firebase/#What_is_Firebase_and_how_it_works)

Reviewer: These should be in the resources/references section

- Developers relied on the platform to synchronize application data in real-time. Lee and Tamplin decided to separate the differentiate real-time architecture from the chat system, which led to the creation of Firebase in 2011. The *Firebase Backend-as-a-Service* platform was unveiled to the public and launched in April 2012.

- The first *Firebase* product launched to hit the market was the **Realtime Database**. This is an API for synchronizing application data between Android synchronization, Web, and iOS devices. Application developers can rely on the platform to create collaborative real-time applications.
- [links](https://www.javatpoint.com/firebase-introduction)

____________________________________________________________________________________________________________________

## **Features:**

![Features](images/features.jpeg)

![Firebase Products & Features](images/features_services.jpg)
[image reference from](https://www.altexsoft.com/blog/firebase-review-pros-cons-alternatives/)

____________________________________________________________________________________________________________________

### **App-Building Services**

> - *Real-time database* - Updates, stores, and sync data between your users. The data is stored as JSON and synchronized with each connected client in real-time. It captures, collects, processes, and enriches incoming data points after they have been made. Data velocity continues to increase and accelerate and now relies more heavily on real-time database technologies. Real-time processing to process handle workloads whose state is constantly changing. Transactions in real-time processing are completed so quickly that the result can be retrieved and processed immediately.

Reviewer: ```suggestion
- *Real-time database* - Updates, stores, and syncs data between your users. The data is stored as JSON and synchronized with each connected client in real time. It captures, collects, processes, and enriches incoming data points after they have been made. Data velocity continues to increase and accelerate and now relies more heavily on real-time database technologies. Real-time processing to process handle workloads whose state is constantly changing. Transactions in real-time processing are completed so quickly that the result can be retrieved and processed immediately.
```

> - *Cloud Firestore* - is a NoSQL document database for the client- and server-side apps document-oriented database. One of the unique features is the syncing of data across multiple client apps using real-time listeners. It supports various file types, numbers, strings, and nested objects. It is secure, reliable, and integrates seamlessly with Firebase and the Google Cloud Platform. It also provides support for offline access, which is especially useful for mobile apps.

Reviewer: ```suggestion
- *Cloud Firestore* - is a NoSQL document database for client and server-side apps. One of the unique features is the syncing of data across multiple client apps using real-time listeners. It supports various file types, numbers, strings, and nested objects. It is secure, reliable, and integrates seamlessly with Firebase and the Google Cloud Platform. It also provides support for offline access, which is especially useful for mobile apps.
```

> - *Cloud Functions* - is a tool for building the mobile app that lets you automatically run backend code in response to trigger events and enables you to firebase features and HTTPS requests. It allows you to write small pieces of JavaScript, which has been the most requested feature.[video-links](https://www.youtube.com/watch?v=QP8sjZuOlFY&t=8s)

Reviewer: ```suggestion
- *Cloud Functions* - is a tool for building mobile apps that lets you automatically run back-end code in response to trigger events and ~~enables you to firebase features and HTTPS requests~~. It allows you to write small JavaScript functions, which has been the most requested feature.[video-links](https://www.youtube.com/watch?v=QP8sjZuOlFY&t=8s)
```

Please rephrase the highlighted section as its meaning is unclear.

> - *Cloud Storage* - is designed for application developers who need to store and serve user-generated content, usually big files like photos or videos. It is a technology that allows storing and managing various media content generated by mobile app users. It is a cloud computing model that holds data on the internet through a cloud computing provider that organizes and operates data storage. We can use Google Cloud Storage to access the duplicate files on the server. It provides agility, durability, and global scale with "anywhere, anytime" data access.[video-links](https://www.youtube.com/watch?v=_tyjqozrEPY&t=5s)

Reviewer: ```suggestion
- *Cloud Storage* - is designed for application developers who need to store and serve user-generated content, usually big files like photos or videos. It is a technology that allows storing and managing various media content generated by mobile app users. It is a cloud computing model that stores data on the internet through a cloud computing provider that organizes and operates data storage. We can use Google Cloud Storage to access the duplicate files on the server. It provides agility, durability, and global scale with "anywhere, anytime" data access.[video-links](https://www.youtube.com/watch?v=_tyjqozrEPY&t=5s)
```

> - *Hosting* - is responsible for providing static and dynamic content and microservices. Can quickly deploy web apps with a single command. It allows us to migrate Firebase hosting content into our domain with SSL security.

Reviewer: ```suggestion
- *Hosting* - is responsible for providing static and dynamic content and microservices. This service can quickly deploy web apps with a single command. It allows us to migrate Firebase hosting content into our domain with SSL security.
```

- *Dynamic Links* - is when users open an app's dynamic link on an android or iOS device but need to install it on their mobile phones, and they will be alerted to install your application. Once the installation completes, the application will start, and the user can access the link. It opens the same active link in the browsers and will redirect them to the website's equivalent content. It can work across application installs.[video-links](https://www.youtube.com/watch?v=LvY1JMcrPF8)

- *Firebase A/B Testing* - allows the app's experience to be easy to run, analyze, and scale product and marketing experiments. It makes you customize your experiment to suit your unique app needs. You determine what elements you want to test, which users to target, what parameters to put in place, and how many variants to include. It allows you to test changes to your app's UI, features, or engagement campaigns.

> - *In-App Messaging* - is a Firebase feature that lets you interact with users actively using your application. You can design your in-app messages to trigger based on your customer’s interests and behaviors. Then, your in-app messaging logic targets users with personalized messages at just the right time. It is more on a technical capability that allows a company to communicate directly with users through the product. It displays lightboxes, tooltips, surveys, banners, and other messaging formats right inside while users engage with it. It tends to have a higher response rate than push notifications or email messages.

Reviewer: ```suggestion
- *In-App Messaging* - is a Firebase feature that lets you interact with users actively using your application. You can design your in-app messages to trigger based on your customer’s interests and behaviors. Then, your in-app messaging logic targets users with personalized messages at just the right time. It is a technical capability that allows a company to communicate directly with users through their product. It displays lightboxes, tooltips, surveys, banners, and other messaging formats right inside while users engage with it. It tends to have a higher response rate than push notifications or email messages.
```

## **Weaknesses:**

![Weakneses](images/weaknesses.jpeg)

- It’s not Open-Source
- Firebase does not work in many countries
- Only NoSQL databases are available
- Slow Queries
- Not all services are free to start
- It’s not cheap and pricing is difficult to forecast
- Runs only on Google Cloud
- Dedicated Servers & Enterprise Support are not an option

____________________________________________________________________________________________________________________

**It’s not Open-Source** - *Firebase users can't modify Firebase source code because it is closed-source. Because of this issue, developers using Firebase end up using alternative platforms*.

**Firebase does not work in many countries** - *Firebase is a subdomain from Google that only works in some countries*.

**Only NoSQL databases are available** - *In Firebase, there is no option to use a relational database. There is a solution for this situation, and a user can craft a complex with Firebase. The implementation may be complex for a beginner using the platform. Firebase uses JSON, and there are virtually no SQL features, so migrating from the database will be more challenging*.

> **Slow Queries** - *Its specific limitations are a downfall of using it, like the frequency limited to 1 per second. No native aggregation queries and the maximum API request size is only 10 MiB. The document size limit is 1 MiB, and the Queries have slow performance*.

Reviewer: ```suggestion
**Slow Queries** - *There are various limitations applied to users depending on their tier, such as the frequency limitation to 1 per second **[ed: on which service?]**. Firebase does not support native aggregation queries and the maximum API request size is only 10 MiB. The document size limit is 1 MiB, and the Queries have slow performance*.
```

This needs to be qualified a little, is it Firestore that is limited in this way?

> **It's not free or cheap, and pricing is complex to forecast** - *Most of its services are available under the Spark Plan, but some are unavailable. Price changes once the user exceeds the free limit. Like the Cloud Vision APIs under the Machine Learning feature, it is only available to Blaze Plan users and charges $1.5/k APIs. Firebase is a feature- high-priced mobile application development platform. It offers a free plan, but that comes with some restrictions. In another way, some open-source technologies deliver more affordable solutions. Estimating Firebase's costs can be difficult for some users, especially when the scalability requirements are unpredictable. The pricing is complicated, and Firebase has specific pricing for every feature. Keeping the costs under a threshold can be challenging with the platform, and there is no easy way to cap or put hard limits on the bill. Understand how the billing works before committing to this model and monitor your costs regularly to avoid unexpected charges by the end of the month*.

Reviewer: ```suggestion
**It's not free or cheap, and pricing is complex to forecast** - *Most of its services are available under the Spark Plan, but some are unavailable. Price changes once the user exceeds the free limit. Like the Cloud Vision APIs under the Machine Learning feature, it is only available to Blaze Plan users and charges $1.5/k API requests. Firebase is a high-priced mobile application development platform. It offers a free plan, but that comes with some restrictions. In another way, some open-source technologies deliver more affordable solutions. 

Estimating Firebase's costs can be difficult for some users, especially when the scalability requirements are unpredictable. The pricing is complicated, and Firebase has specific pricing for every feature. Keeping the costs under a threshold can be challenging with the platform, and there is no easy way to cap or put hard limits on the bill. Understand how the billing works before committing to this model and monitor your costs regularly to avoid unexpected charges by the end of the month*.
```

____________________________________________________________________________________________________________________

## **Alternatives & Comparison:**

- **Parse** is an open source where data migrations are much easier than with Firebase. Parse offers many of the same features and additional benefits. It also runs on any cloud, whereas Firebase only runs on Google Cloud. Users also say data migrations are much more manageable with Parse than with Firebase. Self-hosting is also free.

- **Supabase** is a SQL database rather than NoSQL, offering most features for free.

- **Kuzzle** is another alternative. It is known for improving your TTM (time-to-market) and offers a scalable server, making it an excellent option for fast-growing apps.

- **Back4App** is an alternative to the original Parse hosting service that users can utilize both GraphQL and Rest API. Its live querying functionality makes it easy to execute real-time sync and storage of data.

- [Firebase vs Supabase](https://www.youtube.com/watch?v=yGbGxWMv9KA)

____________________________________________________________________________________________________________________

## **Summary:**

> - What is ***Firebase***? It's Google's all-in-one cloud service, neatly packaged and delivered to developers who want something up and running quickly. No need to worry about provisioning or elastic cloud structures. All you have to do is connect it to your front end, and everything works. It is an impressive technology and a far-reaching platform for mobile-app development that streamlines various tasks and produces quality software solutions. Since Google backs the technology, it tends to be the most robust, scalable, and dynamic framework features of Firebase, giving fierce competition to web service provider companies. This article explained Firebase's advantages, downsides, pricing structure, core features, and best alternative platforms. It also has downsides, like providing NoSQL. Pricing-wise, Firebase operates under two different plans. The first one is Spark, which provides a generous free tier and encompasses most Firebase features. Firebase is fantastic if you want to create something out of nothing in a flash, making it great for rapid prototyping. You can connect to it if you've got the general gist of what you want to do and need a fully configured backend. Then Firebase can be your go-to service.

Reviewer: ```suggestion
- What is ***Firebase***? It's Google's all-in-one cloud service, neatly packaged and delivered to developers who want something up and running quickly. No need to worry about provisioning or elastic cloud structures. All you have to do is connect it to your front end, and everything works. It is an impressive technology and a far-reaching platform for mobile-app development that streamlines various tasks and produces quality software solutions. Since Google backs the technology, it tends to be the most robust, scalable, and dynamic framework features of Firebase, giving fierce competition to web service provider companies. This article explained Firebase's advantages, downsides, pricing structure, core features, and best alternative platforms. It also has downsides, like only providing NoSQL. Pricing-wise, Firebase operates under two different plans. The first one is Spark, which provides a generous free tier and encompasses most Firebase features. Firebase is fantastic if you want to create something out of nothing in a flash, making it great for rapid prototyping. You can connect to it if you've got the general gist of what you want to do and need a fully configured backend. Then Firebase can be your go-to service.
```
